src/view/components/toolbar/motion/Gesture.tsx

```diff
@@ -0,0 +1,71 @@
+import * as React from "react";
+import { CONSTANTS } from "../../../constants";
+import { Dropdown } from "../../Dropdown";
+import SensorButton from "../SensorButton";
+
+const GESTURE_BUTTON_MESSAGE = "Send Gesture";
+interface IProps {
+    gestures: string[];
+    onSelectGestures?: (event: React.ChangeEvent<HTMLSelectElement>) => void;
+    onSendGesture?: (isActive: boolean) => void;
+}
+export class Gesture extends React.Component<IProps> {
+    private sensorButtonRef: React.RefObject<SensorButton> = React.createRef();
+    render() {
+        return (
+            <div
+                className="gesture"
+                style={{
+                    display: "flex",
+                    flexDirection: "column",
+                    justifyContent: "space-between",
+                    height: "75px",
+                }}
+            >
+                <Dropdown
+                    name="gesture selection"
+                    options={this.props.gestures}
+                    onSelect={this.props.onSelectGestures}
+                />
+                <SensorButton
+                    ref={this.sensorButtonRef}
+                    label={GESTURE_BUTTON_MESSAGE}
+                    onMouseDown={() => {
+                        if (this.props.onSendGesture) {
+                            this.props.onSendGesture(true);
+                        }
+                    }}
+                    onMouseUp={() => {
+                        if (this.props.onSendGesture) {
+                            this.props.onSendGesture(false);
+                        }
+                    }}
+                    onKeyDown={this.handleOnKeyDown}
+                    onKeyUp={this.handleOnKeyUp}
+                    type="gesture"
+                />
+            </div>
+        );
+    }
+    private handleOnKeyDown = (e: React.KeyboardEvent) => {
+        if (e.key === CONSTANTS.KEYBOARD_KEYS.ENTER) {
+            this.sensorButtonRef!.current!.setButtonClass(true);
+            if (this.props.onSendGesture) {
+                this.props.onSendGesture(true);
+            }
+        }
+    };
+
+    private handleOnKeyUp = (
+        e: React.KeyboardEvent,
+        onSendGesture?: (isActive: boolean) => void
+    ) => {
+        if (e.key === CONSTANTS.KEYBOARD_KEYS.ENTER) {
+            this.sensorButtonRef!.current!.setButtonClass(false);

```

Comment:
I've never seen this, what does the `!.` operator do?

Reply:
Hey David, this indicated that the value is defined. The reason why it's used is because it's set directly here at line 31 of the same file
                    <SensorButton
                    ref={this.sensorButtonRef}

Reply:
My bad, I don't quite get it. Do you have any reference?

Reply:
A new ! post-fix expression operator may be used to assert that its operand is non-null and non-undefined in contexts where the type checker is unable to conclude that fact. Specifically, the operation x! produces a value of the type of x with null and undefined excluded. Similar to type assertions of the forms <T>x and x as T, the ! non-null assertion operator is simply removed in the emitted JavaScript code. 

https://stackoverflow.com/questions/42273853/in-typescript-what-is-the-exclamation-mark-bang-operator-when-dereferenci

Reply:
Thanks for clarifying!